File: lec0_intro1/ComplexMemory.java
package lec1_intro1;

public class ComplexMemory {
    public static void main(String[] args) {
        // Array of strings - simple and safe!
        String[] words = new String[3];
        
        words[0] = "Java";
        words[1] = "is";
        words[2] = "easier";
        
        // Print them
        for (int i = 0; i < 3; i++) {
            System.out.print(words[i] + " ");
        }
        System.out.println();
        
        // No manual memory management needed!
        // Garbage collector cleans up automatically.
    }
}

/*
COMPLEX DATA IN JAVA:
1. Single, simple reference (String[]) - no pointer confusion.
2. One allocation - Java handles the details.
3. No need to track or free anything.
4. Much harder to make memory management mistakes.
5. Automatic safety checks prevent common errors.

THE BIG PICTURE:
- C gives you complete control but requires careful memory management.
- Java takes care of memory for you, trading a little performance
  for much better safety and simplicity.
- When you're learning data structures, you want to focus on
  the LOGIC, not fighting with malloc/free!
*/

File: lec0_intro1/ComplexMemory.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Imagine we want to store multiple strings
// In C, this gets complicated fast!

int main() {
    // Array of string pointers
    char **words = malloc(3 * sizeof(char *));
    
    if (words == NULL) {
        printf("Memory allocation failed!\n");
        return 1;
    }
    
    // Allocate memory for each string
    words[0] = malloc(10 * sizeof(char));
    words[1] = malloc(10 * sizeof(char));
    words[2] = malloc(10 * sizeof(char));
    
    // Check each allocation (verbose!)
    if (words[0] == NULL || words[1] == NULL || words[2] == NULL) {
        printf("Memory allocation failed!\n");
        // Should free already allocated memory here...
        return 1;
    }
    
    // Copy strings
    strcpy(words[0], "Java");
    strcpy(words[1], "is");
    strcpy(words[2], "easier");
    
    // Print them
    for (int i = 0; i < 3; i++) {
        printf("%s ", words[i]);
    }
    printf("\n");
    
    // Free each string
    for (int i = 0; i < 3; i++) {
        free(words[i]);
    }
    // Free the array of pointers
    free(words);
    
    return 0;
}

/*
COMPLEX DATA IN C:
1. Multiple levels of pointers (char ** for array of strings).
2. Multiple malloc() calls - each can fail!
3. Must track all allocations to free them properly.
4. Easy to make mistakes: memory leaks, double frees, etc.
5. No safety net - errors cause crashes or security bugs.
*/
